lib/trino-hdfs/src/main/java/io/trino/hdfs/util/ConsistentHashingNodeProvider.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.hdfs.util;
+
+import com.google.common.hash.HashFunction;
+import com.google.common.hash.Hashing;
+import com.google.inject.Inject;
+import io.trino.spi.HostAddress;
+import io.trino.spi.Node;
+import io.trino.spi.NodeManager;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class ConsistentHashingNodeProvider
+        implements NodeProvider
+{
+    private final NodeManager nodeManager;
+    // Use murmur3_128 as recommended by Guava for a fast hash that stays stable between runs
+    private final HashFunction hashFunction = Hashing.murmur3_128();
+
+    @Inject
+    public ConsistentHashingNodeProvider(NodeManager nodeManager)
+    {
+        this.nodeManager = requireNonNull(nodeManager, "nodeManager is null");
+    }
+
+    @Override
+    public Optional<HostAddress> getHost(String identifier)
+    {
+        List<HostAddress> currentNodes = nodeManager.getWorkerNodes().stream()
+                .map(Node::getHostAndPort)
+                .sorted(Comparator.comparing(HostAddress::getHostText))
+                .collect(toImmutableList());
+        if (currentNodes.isEmpty()) {
+            return Optional.empty();
+        }
+        int index = Hashing.consistentHash(hashFunction.hashUnencodedChars(identifier), currentNodes.size());
```

Comment:
The current implementation does not truly gives us [consistent hashing](https://dl.acm.org/doi/10.1145/258533.258660). It relies on the ordering of the nodes by `HostAddress`. If a new node is added we may very likely end up adding it to the beginning of the list of nodes. Similarly, if a node is removed, we may end up removing the first node. In both cases, the entire cache is invalidated (e.g., if we have nodes 1, 2, 3, caching A, B, C, respectively, and node 1 is removed, then probably 2 will cache A, 3 will cache B, and C will be distributed among them). We will likely need some memory of the latest applied cache distribution to enable consistent hashing. Perhaps we can extract a subset of https://github.com/Jaskey/ConsistentHash or https://github.com/ishugaliy/allgood-consistent-hash ? [Google has an interesting alternative](https://research.google/pubs/pub45756/) as well.

Reply:
Right, we can refer to https://github.com/qubole/rubix/blob/9b87ec97b2b16e15408aee0dfec54165429fc970/rubix-spi/src/main/java/com/qubole/rubix/spi/ClusterManager.java for an implementation that is more robust to changes in nodes

Reply:
This is fixed in the latest update. We use the same library as Rubix.